src/Utilities/Compiler/Extensions/IOperationExtensions.cs

```diff
@@ -801,6 +801,331 @@ public static bool IsSetMethodInvocation(this IPropertyReferenceOperation operat
+
+            return ValueUsageInfos.Read;
+        }
+
+        public static bool IsInLeftOfDeconstructionAssignment(this IOperation operation, out IDeconstructionAssignmentOperation? deconstructionAssignment)
+        {
+            deconstructionAssignment = null;
+
+            var previousOperation = operation;
+            operation = operation.Parent;
+
+            while (operation != null)
+            {
+                switch (operation.Kind)
+                {
+                    case OperationKind.DeconstructionAssignment:
+                        deconstructionAssignment = (IDeconstructionAssignmentOperation)operation;
+                        return deconstructionAssignment.Target == previousOperation;
+
+                    case OperationKind.Tuple:
+                    case OperationKind.Conversion:
+                    case OperationKind.Parenthesized:
+                        previousOperation = operation;
+                        operation = operation.Parent;
+                        continue;
+
+                    default:
+                        return false;
+                }
+            }
+
+            return false;
+        }
+
+        public static RefKind GetRefKind(this IReturnOperation operation, ISymbol containingSymbol)
+        {
+            var containingMethod = TryGetContainingAnonymousFunctionOrLocalFunction(operation) ?? (containingSymbol as IMethodSymbol);
+            return containingMethod?.RefKind ?? RefKind.None;
+        }
+
+        public static IMethodSymbol? TryGetContainingAnonymousFunctionOrLocalFunction(this IOperation? operation)
+        {
+            operation = operation?.Parent;
+            while (operation != null)
+            {
+                switch (operation.Kind)
+                {
+                    case OperationKind.AnonymousFunction:
+                        return ((IAnonymousFunctionOperation)operation).Symbol;
+
+                    case OperationKind.LocalFunction:
+                        return ((ILocalFunctionOperation)operation).Symbol;
+                }
+
+                operation = operation.Parent;
+            }
+
+            return null;
+        }
+
+        /// <summary>
+        /// Retursn true if the given operation is a regular compound assignment,
+        /// i.e. <see cref="ICompoundAssignmentOperation"/> such as <code>a += b</code>,
+        /// or a special null coalescing compoud assignment, i.e. <see cref="ICoalesceAssignmentOperation"/>
+        /// such as <code>a ??= b</code>.
+        /// </summary>
+        public static bool IsAnyCompoundAssignment(this IOperation operation)
+        {
+            switch (operation)
+            {
+                case ICompoundAssignmentOperation _:
+                case ICoalesceAssignmentOperation _:
+                    return true;
+
+                default:
+                    return false;
+            }
+        }
+#endif
+    }
+
+    [Flags]
+    internal enum ValueUsageInfos
```

Comment:
Nit: Can we have the name match the Roslyn name, i.e. `ValueUsageInfo`? Also please move it to a separate file outside Extensions folder.

Reply:
There is this warning `Flags enums should have plural names`, i will suppress it then

Reply:
Seems like a false positive. I think `Info` means plural, though probably it is not a legitimate word so maybe its an acceptable false positive :-)? Suppressing should be fine.